File: src/types/video.ts
// Video information returned by yt-dlp --dump-json
export interface VideoInfo {
  videoId: string;
  title: string;
  uploader: string;
  duration: number; // seconds
  thumbnail: string;
  description: string;
  formats: VideoFormat[];
  availableQualities: string[]; // e.g. ['2160p', '1080p', '720p', '480p']
  subtitles: Record<string, SubtitleTrack[]>;
  automaticCaptions: Record<string, SubtitleTrack[]>;
}

export interface VideoFormat {
  formatId: string;
  ext: string;
  quality: string;
  filesize?: number;
  vcodec?: string;
  acodec?: string;
  height?: number;
  width?: number;
  tbr?: number; // total bitrate
}

export interface SubtitleTrack {
  ext: string;
  url?: string;
  name?: string;
}

export interface SubtitleOptions {
  enabled: boolean;
  language: string; // BCP-47 code, e.g. 'en', 'hi', or 'all'
  mode: 'embed' | 'sidecar';
  includeAuto: boolean;
}

// Playlist item from yt-dlp --flat-playlist
export interface PlaylistItem {
  id: string;
  title: string;
  url: string;
  duration?: number;
  thumbnail?: string;
  uploader?: string;
}

export interface PlaylistInfo {
  id: string;
  title: string;
  uploader: string;
  thumbnail?: string;
  entryCount: number;
  entries: PlaylistItem[];
}

File: src-tauri/src/commands/video.rs
use crate::security::url_validator::validate_url;
use crate::services::ytdlp::info::{fetch_playlist_info, fetch_video_info};
use anyhow::Result;
use serde::{Deserialize, Serialize};

// ─── Types mirroring the frontend TypeScript types ────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VideoFormat {
    pub format_id: String,
    pub ext: String,
    pub quality: String,
    pub filesize: Option<u64>,
    pub vcodec: Option<String>,
    pub acodec: Option<String>,
    pub height: Option<u32>,
    pub width: Option<u32>,
    pub tbr: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubtitleTrack {
    pub ext: String,
    pub url: Option<String>,
    pub name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VideoInfo {
    pub video_id: String,
    pub title: String,
    pub uploader: String,
    pub duration: u64,
    pub thumbnail: String,
    pub description: String,
    pub formats: Vec<VideoFormat>,
    pub available_qualities: Vec<String>,
    pub subtitles: std::collections::HashMap<String, Vec<SubtitleTrack>>,
    pub automatic_captions: std::collections::HashMap<String, Vec<SubtitleTrack>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaylistItem {
    pub id: String,
    pub title: String,
    pub url: String,
    pub duration: Option<u64>,
    pub thumbnail: Option<String>,
    pub uploader: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaylistInfo {
    pub id: String,
    pub title: String,
    pub uploader: String,
    pub thumbnail: Option<String>,
    pub entry_count: usize,
    pub entries: Vec<PlaylistItem>,
}

// ─── Commands ─────────────────────────────────────────────────────────────────

#[tauri::command]
pub async fn get_video_info(url: String) -> Result<VideoInfo, String> {
    // Validate URL first (security)
    let validated_url = validate_url(&url).map_err(|e| e.to_string())?;

    let info = fetch_video_info(&validated_url)
        .await
        .map_err(|e| format!("Failed to fetch video info: {}", e))?;

    Ok(info)
}

#[tauri::command]
pub async fn get_playlist_info(url: String) -> Result<PlaylistInfo, String> {
    let validated_url = validate_url(&url).map_err(|e| e.to_string())?;

    let info = fetch_playlist_info(&validated_url)
        .await
        .map_err(|e| format!("Failed to fetch playlist: {}", e))?;

    Ok(info)
}
